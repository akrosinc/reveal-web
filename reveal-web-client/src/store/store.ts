import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import userReducer from '../features/reducers/user';
import toastReducer from '../features/reducers/tostify';

export const store = configureStore({
  reducer: {
    user: userReducer,
    toast: toastReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
