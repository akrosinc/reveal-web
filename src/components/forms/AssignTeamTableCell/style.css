.assignment-cell .assignment-label {
  text-overflow: ellipsis;
  white-space: nowrap;
  max-width: calc(100% - 120px);
  display: inline-block;
  overflow: hidden;
}
