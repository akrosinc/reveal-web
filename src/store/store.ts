import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import userReducer from '../features/reducers/user';
import loaderReducer from '../features/reducers/loader';

export const store = configureStore({
  reducer: {
    user: userReducer,
    loader: loaderReducer
  }
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
