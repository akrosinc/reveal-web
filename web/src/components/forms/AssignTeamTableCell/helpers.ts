/** geting the button id for the AssignTeamButton
 * @param {string} jurisdictionId the id of the jurisdiction relevant to the table row
 * @returns {string} the id of the TeamAssignmentButton and target of the TeamAssignmentPopover
 */
export const getButtonId = (jurisdictionId: string): string => `plan-assignment-${jurisdictionId}`;

/** geting the button id of the Team Assignment form
 * @param {string} jurisdictionId the id of the jurisdiction relevant to the table row
 * @returns {string} the name of the form containing the OrganizationSelect component
 */
export const getFormName = (jurisdictionId: string): string =>
  `plan-assignment-form-${jurisdictionId}`;
