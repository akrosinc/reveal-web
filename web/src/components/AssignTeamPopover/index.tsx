import React, { MouseEvent } from 'react';
import { Button, Form, FormGroup, Popover, PopoverBody, PopoverHeader } from 'reactstrap';
import { CLEAR, NO_TEAMS_LOADED_MESSAGE, SAVE, SELECT_TEAMS_TO_ASSIGN } from '../../configs/lang';
import { stopPropagationAndPreventDefault } from '../../helpers/utils';
import { Organization } from '../../store/ducks/opensrp/organizations';
import OrganizationSelect from '../forms/OrganizationSelect';

/** Interface for AssignTeamPopover */
export interface AssignTeamPopoverProps {
  formName: string; // id of the continer form element for the nested ReactSelect el
  isActive: boolean; // whether or not the popover should be visible
  jurisdictionId: string; // the id of the relevant Jurisdiction
  onClearAssignmentsButtonClick: (e: MouseEvent) => void; // what to do when clicking 'Clear'
  onSaveAssignmentsButtonClick: (e: MouseEvent) => void; // what to do when clicking 'Save'
  onToggle: () => void; // what to do when the popover toggles
  organizationsById: { [key: string]: Organization } | null; // all the organizations
  planId: string; // the id of the current Plan
  target: string; // the id of the adjacent button which toggles the popup
  parentIds?: string[];
  parentAssignments?: string[];
}

/** default Component for AssignTeamPopover */
const AssignTeamPopover = (props: AssignTeamPopoverProps) => {
  const {
    formName,
    isActive,
    jurisdictionId,
    onClearAssignmentsButtonClick,
    onSaveAssignmentsButtonClick,
    onToggle,
    organizationsById,
    planId,
    target,
    parentIds,
    parentAssignments,
  } = props;

  // define props for the child Organization Select component
  const organizationSelectProps = {
    jurisdictionId,
    name: formName,
    parentAssignments,
    parentIds,
    planId,
  };

  // define props for the actual popover component
  const popoverProps = {
    isOpen: isActive,
    onClick: (e: React.MouseEvent) => stopPropagationAndPreventDefault(e),
    target,
    toggle: () => onToggle(),
  };

  // hide clear all button if we have parent assignments
  const hasParentAssignments = parentAssignments ? parentAssignments.length > 0 : false;

  return (
    <Popover {...popoverProps} style={{ minWidth: '16rem' }}>
      <PopoverHeader>{SELECT_TEAMS_TO_ASSIGN}</PopoverHeader>
      <PopoverBody>
        {organizationsById ? (
          <Form name={formName}>
            <FormGroup>
              <OrganizationSelect {...organizationSelectProps} />
            </FormGroup>
            <FormGroup>
              {!hasParentAssignments && (
                <Button
                  color="default"
                  onClick={onClearAssignmentsButtonClick}
                  size="xs"
                  outline={true}
                >
                  {CLEAR}
                </Button>
              )}
              <Button color="primary" onClick={onSaveAssignmentsButtonClick} size="xs">
                {SAVE}
              </Button>
            </FormGroup>
          </Form>
        ) : (
          <p>{NO_TEAMS_LOADED_MESSAGE}</p>
        )}
      </PopoverBody>
    </Popover>
  );
};

export default AssignTeamPopover;
