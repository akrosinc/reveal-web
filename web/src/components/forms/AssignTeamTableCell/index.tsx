import reducerRegistry from '@onaio/redux-reducer-registry';
import React, { MouseEvent, useState } from 'react';
import { connect } from 'react-redux';
import { Store } from 'redux';
import { TEAMS_ASSIGNED } from '../../../configs/lang';
import { stopPropagationAndPreventDefault } from '../../../helpers/utils';
import assignmentReducer, {
  Assignment,
  fetchAssignments,
  getAssignmentsArrayByPlanId,
  reducerName as assignmentReducerName,
  resetPlanAssignments,
} from '../../../store/ducks/opensrp/assignments';
import organizationsReducer, {
  getOrganizationsById,
  Organization,
  reducerName as organizationsReducerName,
} from '../../../store/ducks/opensrp/organizations';
import AssignTeamButton, { AssignTeamButtonProps } from '../../AssignTeamButton';
import AssignTeamPopover, { AssignTeamPopoverProps } from '../../AssignTeamPopover';
import { getButtonId, getFormName } from './helpers';
import './style.css';

reducerRegistry.register(assignmentReducerName, assignmentReducer);
reducerRegistry.register(organizationsReducerName, organizationsReducer);

/** Interface for Assign Teams cell props */
export interface AssignTeamCellProps {
  assignments: Assignment[]; // all the assignments for the plan-jurisdiction
  assignmentsArray: Assignment[]; // all the assignments for the plan
  assignButton?: React.ElementType;
  assignButtonProps?: AssignTeamButtonProps;
  assignPopover?: React.ElementType;
  assignPopoverProps?: AssignTeamPopoverProps;
  resetPlanAssignmentsAction: typeof resetPlanAssignments;
  jurisdictionId: string;
  organizationsById: { [key: string]: Organization } | null;
  planId: string;
  parentIds?: string[];
  parentAssignments?: string[];
}

/** Component that will be rendered in IRS planning table cells
 * showing the button and card in the Assign Teams column
 */
const AssignTeamTableCell = (props: AssignTeamCellProps) => {
  const {
    assignments,
    assignmentsArray,
    assignButton,
    assignPopover,
    resetPlanAssignmentsAction,
    jurisdictionId,
    organizationsById,
    planId,
    parentIds,
    parentAssignments,
  } = props;
  const [isActive, setIsActive] = useState<boolean>(false);

  // toggle isActive state on button click
  const onPlanAssignmentButtonClick = (e: MouseEvent) => {
    stopPropagationAndPreventDefault(e);
    setIsActive(!isActive);
  };

  // remove assignments for this jurisdicition and save to state,
  // and toggle isActive state on clear button click
  const onClearAssignmentsButtonClick = (e: MouseEvent) => {
    stopPropagationAndPreventDefault(e);
    const nextAssignments = assignmentsArray.filter(
      (a: Assignment) => a.jurisdiction !== jurisdictionId
    );
    resetPlanAssignmentsAction({ [planId]: [...nextAssignments] });
    setIsActive(!isActive);
  };

  // toggle isActive state when clicking save button
  const onSaveAssignmentsButtonClick = (e: MouseEvent) => {
    stopPropagationAndPreventDefault(e);
    setIsActive(!isActive);
  };

  // define the props being passed to the default AssignTeamButton
  const assignTeamButtonProps: AssignTeamButtonProps = props.assignButtonProps || {
    id: getButtonId(jurisdictionId),
    onClick: onPlanAssignmentButtonClick,
  };

  // define the props for the default AssignTeamPopover
  const assignTeamPopoverProps: AssignTeamPopoverProps = props.assignPopoverProps || {
    formName: getFormName(jurisdictionId),
    isActive,
    jurisdictionId,
    onClearAssignmentsButtonClick,
    onSaveAssignmentsButtonClick,
    onToggle: () => setIsActive(!isActive),
    organizationsById,
    parentAssignments,
    parentIds,
    planId,
    target: getButtonId(jurisdictionId),
  };

  const assignedOrganizations = assignments
    .map(
      (a: Assignment) =>
        organizationsById &&
        organizationsById[a.organization] &&
        organizationsById[a.organization].name
    )
    .filter(org => !!org)
    .join(', ');

  return (
    <div onClick={stopPropagationAndPreventDefault} className="assignment-cell">
      <span className="assignment-label" title={`${assignments.length} ${TEAMS_ASSIGNED}`}>
        {assignedOrganizations}
      </span>
      {assignButton || <AssignTeamButton {...assignTeamButtonProps} />}
      {assignPopover || <AssignTeamPopover {...assignTeamPopoverProps} />}
    </div>
  );
};

export { AssignTeamTableCell };

/** map state to props
 * @param {partial<store>} - the redux store
 * @param {AssignTeamCellProps} ownProps - the props
 *
 * @returns {AssignTeamCellProps} - ownProps and organizationsById
 */
const mapStateToProps = (state: Partial<Store>, ownProps: any): AssignTeamCellProps => {
  const organizationsById = getOrganizationsById(state);
  const assignmentsArray = getAssignmentsArrayByPlanId(state, ownProps.planId);
  const idsToCheckAssignments =
    ownProps.parentIds && ownProps.parentIds.length
      ? [ownProps.jurisdictionId, ...ownProps.parentIds]
      : [ownProps.jurisdictionId];
  const parentAssignments: string[] = [];
  const uniqueOptionsIds: string[] = [];
  const assignments: Assignment[] = [];
  assignmentsArray.forEach((a: Assignment) => {
    if (idsToCheckAssignments.includes(a.jurisdiction)) {
      if (a.jurisdiction !== ownProps.jurisdictionId) {
        parentAssignments.push(a.organization);
      }
      if (!uniqueOptionsIds.includes(a.organization)) {
        uniqueOptionsIds.push(a.organization);
        assignments.push(a);
      }
    }
  });

  return {
    ...ownProps,
    assignments,
    assignmentsArray,
    organizationsById,
    parentAssignments,
  } as AssignTeamCellProps;
};

/** map props to actions that may be dispatched by component */
const mapDispatchToProps = {
  fetchAssignmentsActionCreator: fetchAssignments,
  resetPlanAssignmentsAction: resetPlanAssignments,
};

/** Create Connected AssignTeamTableCell */
const ConnectedAssignTeamTableCell = connect(
  mapStateToProps,
  mapDispatchToProps
)(AssignTeamTableCell);

export default ConnectedAssignTeamTableCell;
