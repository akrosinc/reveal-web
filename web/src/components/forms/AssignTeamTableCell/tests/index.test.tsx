import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Provider } from 'react-redux';
import AssignTeamTableCell from '..';
import store from '../../../../store';
import assignmentReducer, {
  fetchAssignments,
  reducerName as assignmentReducerName,
} from '../../../../store/ducks/opensrp/assignments';
import organizationsReducer, {
  fetchOrganizations,
  reducerName as organizationsReducerName,
} from '../../../../store/ducks/opensrp/organizations';
import * as fixtures from '../../../../store/ducks/tests/fixtures';
import { AssignTeamPopoverProps } from '../../../AssignTeamPopover';

jest.mock('../../../../configs/env');

reducerRegistry.register(assignmentReducerName, assignmentReducer);
reducerRegistry.register(organizationsReducerName, organizationsReducer);

describe('/containers/forms/AssignTeamTableCell', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(<AssignTeamTableCell />);
  });

  it('renders correctly', () => {
    const props = {
      assignPopover: <div id="custom-popover" />,
      jurisdictionId: 'outpost-number-one',
      planId: 'alpha',
    };

    const wrapper = mount(
      <Provider store={store}>
        <AssignTeamTableCell {...props} />
      </Provider>
    );
    expect(toJson(wrapper.find('button#plan-assignment-outpost-number-one'))).toMatchSnapshot(
      'AssignTeamButton'
    );
    expect(toJson(wrapper.find('span.assignment-label'))).toMatchSnapshot('Team Count Text');

    store.dispatch(fetchAssignments(fixtures.assignments));
    wrapper.update();
    expect(toJson(wrapper.find('span.assignment-label'))).toMatchSnapshot('Updated teams');
  });

  it('should toggle teamPopOver correctly', () => {
    const btnId = 'plan-assignment-outpost-number-one';
    const div = document.createElement('div');
    div.setAttribute('id', btnId);
    document.body.appendChild(div);

    store.dispatch(fetchOrganizations(fixtures.organizations));
    store.dispatch(fetchAssignments(fixtures.assignments));

    const props = {
      jurisdictionId: 'outpost-number-one',
      planId: 'alpha',
    };
    const wrapper = mount(
      <Provider store={store}>
        <AssignTeamTableCell {...props} />
      </Provider>
    );

    expect(wrapper.find('AssignTeamPopover').length).toEqual(1);

    let assignTeamPopoverProps = wrapper
      .find('AssignTeamPopover')
      .props() as AssignTeamPopoverProps;
    expect(assignTeamPopoverProps.isActive).toBeFalsy();

    // toggles assignTeamPopover
    wrapper.find(`button#${btnId}`).simulate('click');
    assignTeamPopoverProps = wrapper.find('AssignTeamPopover').props() as AssignTeamPopoverProps;
    expect(assignTeamPopoverProps.isActive).toBeTruthy();

    // save button
    expect(
      wrapper
        .find('Button')
        .at(2)
        .text()
    ).toEqual('Save');
    wrapper
      .find('Button')
      .at(2)
      .simulate('click');
    assignTeamPopoverProps = wrapper.find('AssignTeamPopover').props() as AssignTeamPopoverProps;
    expect(assignTeamPopoverProps.isActive).toBeFalsy();

    // open assignTeamPopover
    wrapper.find(`button#${btnId}`).simulate('click');

    // clear button
    expect(
      wrapper
        .find('Button')
        .at(1)
        .text()
    ).toEqual('Clear');
    wrapper
      .find('Button')
      .at(1)
      .simulate('click');
    assignTeamPopoverProps = wrapper.find('AssignTeamPopover').props() as AssignTeamPopoverProps;
    expect(assignTeamPopoverProps.isActive).toBeFalsy();
  });

  it('renders correctly with parent assignments', () => {
    store.dispatch(fetchOrganizations(fixtures.organizations));
    store.dispatch(fetchAssignments(fixtures.assignments));
    const props = {
      assignPopover: <div id="custom-popover" />,
      jurisdictionId: 'outpost-number-one',
      parentIds: ['blue-base'],
      planId: 'alpha',
    };

    const wrapper = mount(
      <Provider store={store}>
        <AssignTeamTableCell {...props} />
      </Provider>
    );

    expect(toJson(wrapper.find('button#plan-assignment-outpost-number-one'))).toMatchSnapshot(
      'AssignTeamButton for outpost-number-one assignment'
    );

    expect(toJson(wrapper.find('span.assignment-label'))).toMatchSnapshot(
      'Team Count Text with parent assignments'
    );

    store.dispatch(fetchAssignments(fixtures.assignments));
    wrapper.update();
    expect(toJson(wrapper.find('span.assignment-label'))).toMatchSnapshot(
      'Updated teams with parent assignments'
    );
  });
});
