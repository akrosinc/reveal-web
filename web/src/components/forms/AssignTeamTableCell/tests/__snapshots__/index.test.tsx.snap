// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/containers/forms/AssignTeamTableCell renders correctly with parent assignments: AssignTeamButton for outpost-number-one assignment 1`] = `
<button
  aria-label={null}
  className="assign-team-button btn btn-primary btn-sm"
  id="plan-assignment-outpost-number-one"
  onClick={[Function]}
  style={
    Object {
      "float": "right",
    }
  }
  type="button"
>
  Assign Teams
</button>
`;

exports[`/containers/forms/AssignTeamTableCell renders correctly with parent assignments: Team Count Text with parent assignments 1`] = `
<span
  className="assignment-label"
  title="4 Teams Assigned"
/>
`;

exports[`/containers/forms/AssignTeamTableCell renders correctly with parent assignments: Updated teams with parent assignments 1`] = `
<span
  className="assignment-label"
  title="4 Teams Assigned"
/>
`;

exports[`/containers/forms/AssignTeamTableCell renders correctly: AssignTeamButton 1`] = `
<button
  aria-label={null}
  className="assign-team-button btn btn-primary btn-sm"
  id="plan-assignment-outpost-number-one"
  onClick={[Function]}
  style={
    Object {
      "float": "right",
    }
  }
  type="button"
>
  Assign Teams
</button>
`;

exports[`/containers/forms/AssignTeamTableCell renders correctly: Team Count Text 1`] = `
<span
  className="assignment-label"
  title="0 Teams Assigned"
/>
`;

exports[`/containers/forms/AssignTeamTableCell renders correctly: Updated teams 1`] = `
<span
  className="assignment-label"
  title="2 Teams Assigned"
/>
`;
